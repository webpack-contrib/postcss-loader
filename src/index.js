import path from "path";

import postcssPackage from "postcss/package.json";

import schema from "./options.json";
import {
  loadConfig,
  getPostcssOptions,
  exec,
  normalizeSourceMap,
  normalizeSourceMapAfterPostcss,
  findPackageJSONDir,
  getPostcssImplementation,
  reportError,
  warningFactory,
} from "./utils";

let hasExplicitDependencyOnPostCSS = false;

/**
 * **PostCSS Loader**
 *
 * Loads && processes CSS with [PostCSS](https://github.com/postcss/postcss)
 *
 * @method loader
 *
 * @param {String} content Source
 * @param {Object} sourceMap Source Map
 * @param {Object} meta Meta
 *
 * @return {callback} callback Result
 */
export default async function loader(content, sourceMap, meta) {
  const options = this.getOptions(schema);
  const callback = this.async();
  const configOption =
    typeof options.postcssOptions === "undefined" ||
    typeof options.postcssOptions.config === "undefined"
      ? true
      : options.postcssOptions.config;

  let implementation;

  try {
    implementation = getPostcssImplementation(this, options.implementation);
  } catch (error) {
    callback(error);

    return;
  }

  if (!implementation) {
    callback(
      new Error(
        `The Postcss implementation "${options.implementation}" not found`
      )
    );

    return;
  }

  let loadedConfig;

  if (configOption) {
    try {
      loadedConfig = await loadConfig(
        this,
        configOption,
        options.postcssOptions
      );
    } catch (error) {
      callback(error);

      return;
    }
  }

  const { plugins, processOptions } = await getPostcssOptions(
    this,
    loadedConfig,
    options.postcssOptions
  );

  const useSourceMap =
    typeof options.sourceMap !== "undefined"
      ? options.sourceMap
      : this.sourceMap;

  if (useSourceMap) {
    processOptions.map = {
      inline: false,
      annotation: false,
      ...processOptions.map,
    };
  }

  if (sourceMap && processOptions.map) {
    processOptions.map.prev = normalizeSourceMap(sourceMap, this.context);
  }

  let root;

  // Reuse PostCSS AST from other loaders
  if (
    meta &&
    meta.ast &&
    meta.ast.type === "postcss" &&
    // eslint-disable-next-line global-require
    require("semver").satisfies(meta.ast.version, `^${postcssPackage.version}`)
  ) {
    ({ root } = meta.ast);
  }

  if (!root && options.execute) {
    // eslint-disable-next-line no-param-reassign
    content = exec(content, this);
  }

  let result;
  let processor;

  try {
    processor = implementation(plugins);
    result = await processor.process(root || content, processOptions);
  } catch (error) {
    // Check postcss versions to avoid using PostCSS 7.
    // For caching reasons, we use the readFileSync and existsSync functions from the context,
    // not the functions from the `fs` module.
    if (
      !hasExplicitDependencyOnPostCSS &&
      processor &&
      processor.version &&
      processor.version.startsWith("7.")
    ) {
      // The `findPackageJsonDir` function returns `string` or `null`.
      // This is used to do for caching, that is, an explicit comparison with `undefined`
      // is used to make the condition body run once.
      const packageJSONDir = findPackageJSONDir(
        process.cwd(),
        this.fs.statSync
      );

      if (packageJSONDir) {
        let bufferOfPackageJSON;

        try {
          bufferOfPackageJSON = this.fs.readFileSync(
            path.resolve(packageJSONDir, "package.json"),
            "utf8"
          );
        } catch (_error) {
          // Nothing
        }

        if (bufferOfPackageJSON) {
          let pkg;

          try {
            pkg = JSON.parse(bufferOfPackageJSON);
          } catch (_error) {
            // Nothing
          }

          if (pkg) {
            if (!pkg.dependencies.postcss && !pkg.devDependencies.postcss) {
              this.emitWarning(
                new Error(
                  "Add postcss as project dependency. postcss is not a peer dependency for postcss-loader. " +
                    "Use `npm install postcss` or `yarn add postcss`"
                )
              );
            } else {
              hasExplicitDependencyOnPostCSS = true;
            }
          }
        }
      }
    }

    reportError(this, callback, error);

    return;
  }

  for (const warning of result.warnings()) {
    this.emitWarning(warningFactory(warning));
  }

  for (const message of result.messages) {
    // eslint-disable-next-line default-case
    switch (message.type) {
      case "dependency":
        this.addDependency(message.file);
        break;
      case "build-dependency":
        this.addBuildDependency(message.file);
        break;
      case "missing-dependency":
        this.addMissingDependency(message.file);
        break;
      case "context-dependency":
        this.addContextDependency(message.file);
        break;
      case "dir-dependency":
        this.addContextDependency(message.dir);
        break;
      case "asset":
        if (message.content && message.file) {
          this.emitFile(
            message.file,
            message.content,
            message.sourceMap,
            message.info
          );
        }
    }
  }

  // eslint-disable-next-line no-undefined
  let map = result.map ? result.map.toJSON() : undefined;

  if (map && useSourceMap) {
    map = normalizeSourceMapAfterPostcss(map, this.context);
  }

  let ast;

  try {
    ast = {
      type: "postcss",
      version: result.processor.version,
      root: result.root,
    };
  } catch (error) {
    reportError(this, callback, error);

    return;
  }

  callback(null, result.css, map, { ast });
}
