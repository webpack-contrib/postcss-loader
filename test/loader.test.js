import path from "path";

import postcss from "postcss";

// eslint-disable-next-line import/no-namespace
import * as utils from "../src/utils";

import {
  compile,
  getCompiler,
  getErrors,
  getCodeFromBundle,
  getWarnings,
} from "./helpers/index";

describe("loader", () => {
  it("should work", async () => {
    const compiler = getCompiler("./css/index.js");
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle("style.css", stats);

    expect(codeFromBundle.css).toMatchSnapshot("css");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it("should throw an error on invalid syntax", async () => {
    const compiler = getCompiler("./css/index.js", {
      postcssOptions: {
        hideNothingWarning: true,
        parser: "sugarss",
      },
    });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should emit warning using the "messages" API', async () => {
    const plugin = () => (css, result) => {
      css.walkDecls((node) => {
        node.warn(result, "<Message>");
      });
    };

    const postcssPlugin = postcss.plugin("postcss-plugin", plugin);

    const compiler = getCompiler("./css/index.js", {
      postcssOptions: {
        plugins: [postcssPlugin()],
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle("style.css", stats);

    expect(codeFromBundle.css).toMatchSnapshot("css");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should emit asset using the "messages" API', async () => {
    const plugin = () => (css, result) => {
      result.messages.push({
        type: "asset",
        file: "sprite.svg",
        content: "<svg>...</svg>",
        plugin,
      });
    };

    const postcssPlugin = postcss.plugin("postcss-assets", plugin);
    const compiler = getCompiler("./css/index.js", {
      postcssOptions: {
        plugins: [postcssPlugin()],
      },
    });
    const stats = await compile(compiler);

    // eslint-disable-next-line no-underscore-dangle
    expect(stats.compilation.assets["sprite.svg"]).toBeDefined();
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it('should register dependencies using the "messages" API', async () => {
    const plugin = () => (css, result) => {
      result.messages.push(
        {
          type: "build-dependency",
          file: path.resolve(__dirname, "fixtures", "build-dep.html"),
          content: "",
          plugin,
        },
        {
          type: "missing-dependency",
          file: path.resolve(__dirname, "fixtures", "missing-dep.html"),
          content: "",
          plugin,
        },
        {
          type: "context-dependency",
          file: path.resolve(__dirname, "fixtures", "deps"),
          content: "",
          plugin,
        },
        {
          type: "dir-dependency",
          dir: path.resolve(__dirname, "fixtures", "deps2"),
          content: "",
          plugin,
        }
      );
    };

    const postcssPlugin = postcss.plugin("postcss-plugin", plugin);
    const compiler = getCompiler("./css/index.js", {
      postcssOptions: {
        plugins: [postcssPlugin()],
      },
    });

    const stats = await compile(compiler);
    const { contextDependencies, missingDependencies, buildDependencies } =
      stats.compilation;

    expect(contextDependencies).toContain(
      path.resolve(__dirname, "fixtures", "deps")
    );
    expect(contextDependencies).toContain(
      path.resolve(__dirname, "fixtures", "deps2")
    );
    expect(missingDependencies).toContain(
      path.resolve(__dirname, "fixtures", "missing-dep.html")
    );
    expect(buildDependencies).toContain(
      path.resolve(__dirname, "fixtures", "build-dep.html")
    );

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it("should reuse PostCSS AST", async () => {
    const spy = jest.fn();
    const compiler = getCompiler(
      "./css/index.js",
      {},
      {
        module: {
          rules: [
            {
              test: /\.(css|sss)$/i,
              use: [
                {
                  loader: require.resolve("./helpers/testLoader"),
                  options: {},
                },
                {
                  loader: path.resolve(__dirname, "../src"),
                  options: {
                    postcssOptions: { hideNothingWarning: true },
                  },
                },
                {
                  loader: require.resolve("./helpers/astLoader"),
                  options: { spy },
                },
              ],
            },
          ],
        },
      }
    );
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle("style.css", stats);

    expect(spy).toHaveBeenCalledTimes(1);
    expect(codeFromBundle.css).toMatchSnapshot("css");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });

  it("should work with SugarSS", async () => {
    const compiler = getCompiler("./sss/index.js", {
      postcssOptions: {
        parser: "sugarss",
        hideNothingWarning: true,
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle("style.sss", stats);

    expect(codeFromBundle.css).toMatchSnapshot("css");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats)).toMatchSnapshot("errors");
  });
});

describe("check postcss versions to avoid using PostCSS 7", () => {
  it("should emit a warning if postcss version is not explicitly specified when the loader is failed", async () => {
    const spy = jest
      .spyOn(utils, "findPackageJSONDir")
      .mockReturnValue(
        path.resolve(__dirname, "./fixtures/package-json-files/no-postcss")
      );

    const compiler = getCompiler("./css/index.js", {
      implementation: (...args) => {
        const result = postcss(...args);

        result.version = "7.0.0";
        result.process = () =>
          Promise.reject(new Error("Something went wrong."));

        return result;
      },
    });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats, true)).toMatchSnapshot("errors");

    spy.mockRestore();
  });

  it("should not show a warning if 'postcss' version is explicitly defined in 'dependencies'", async () => {
    const spy = jest
      .spyOn(utils, "findPackageJSONDir")
      .mockReturnValue(
        path.resolve(
          __dirname,
          "./fixtures/package-json-files/postcss-v8-in-dependencies"
        )
      );

    const compiler = getCompiler("./css/index.js", {
      implementation: (...args) => {
        const result = postcss(...args);

        result.version = "7.0.0";
        result.process = () =>
          Promise.reject(new Error("Something went wrong."));

        return result;
      },
    });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats, true)).toMatchSnapshot("errors");

    spy.mockRestore();
  });

  it("should not show a warning if 'postcss' version is explicitly defined in 'devDependencies'", async () => {
    const spy = jest
      .spyOn(utils, "findPackageJSONDir")
      .mockReturnValue(
        path.resolve(
          __dirname,
          "./fixtures/package-json-files/postcss-v8-in-devDependencies"
        )
      );

    const compiler = getCompiler("./css/index.js", {
      implementation: (...args) => {
        const result = postcss(...args);

        result.version = "7.0.0";
        result.process = () =>
          Promise.reject(new Error("Something went wrong."));

        return result;
      },
    });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats, true)).toMatchSnapshot("errors");

    spy.mockRestore();
  });

  it("should not show a warning if 'package.json' file was not found", async () => {
    const spy = jest
      .spyOn(utils, "findPackageJSONDir")
      .mockReturnValue(
        path.resolve(__dirname, "./fixtures/package-json-files/unknown")
      );

    const compiler = getCompiler("./css/index.js", {
      implementation: (...args) => {
        const result = postcss(...args);

        result.version = "7.0.0";
        result.process = () =>
          Promise.reject(new Error("Something went wrong."));

        return result;
      },
    });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats, true)).toMatchSnapshot("errors");

    spy.mockRestore();
  });

  it("should work when 'processor' throw an error", async () => {
    const spy = jest
      .spyOn(utils, "findPackageJSONDir")
      .mockReturnValue(
        path.resolve(__dirname, "./fixtures/package-json-files/no-postcss")
      );

    const compiler = getCompiler("./css/index.js", {
      implementation: () => {
        throw new Error("Error in implementation");
      },
    });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot("warnings");
    expect(getErrors(stats, true)).toMatchSnapshot("errors");

    spy.mockRestore();
  });
});
