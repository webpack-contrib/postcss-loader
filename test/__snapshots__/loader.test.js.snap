// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check postcss versions to avoid using PostCSS 7 should emit a warning if postcss version is not explicitly specified when the loader is failed: errors 1`] = `
[
  "ModuleBuildError: Module build failed (from \`replaced original path\`):
Error: Something went wrong.",
]
`;

exports[`check postcss versions to avoid using PostCSS 7 should emit a warning if postcss version is not explicitly specified when the loader is failed: warnings 1`] = `
[
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Add postcss as project dependency. postcss is not a peer dependency for postcss-loader. Use \`npm install postcss\` or \`yarn add postcss\`",
]
`;

exports[`check postcss versions to avoid using PostCSS 7 should not show a warning if 'package.json' file was not found: errors 1`] = `
[
  "ModuleBuildError: Module build failed (from \`replaced original path\`):
Error: Something went wrong.",
]
`;

exports[`check postcss versions to avoid using PostCSS 7 should not show a warning if 'package.json' file was not found: warnings 1`] = `[]`;

exports[`check postcss versions to avoid using PostCSS 7 should not show a warning if 'postcss' version is explicitly defined in 'dependencies': errors 1`] = `
[
  "ModuleBuildError: Module build failed (from \`replaced original path\`):
Error: Something went wrong.",
]
`;

exports[`check postcss versions to avoid using PostCSS 7 should not show a warning if 'postcss' version is explicitly defined in 'dependencies': warnings 1`] = `[]`;

exports[`check postcss versions to avoid using PostCSS 7 should not show a warning if 'postcss' version is explicitly defined in 'devDependencies': errors 1`] = `
[
  "ModuleBuildError: Module build failed (from \`replaced original path\`):
Error: Something went wrong.",
]
`;

exports[`check postcss versions to avoid using PostCSS 7 should not show a warning if 'postcss' version is explicitly defined in 'devDependencies': warnings 1`] = `[]`;

exports[`check postcss versions to avoid using PostCSS 7 should work when 'processor' throw an error: errors 1`] = `
[
  "ModuleBuildError: Module build failed (from \`replaced original path\`):
Error: Error in implementation",
]
`;

exports[`check postcss versions to avoid using PostCSS 7 should work when 'processor' throw an error: warnings 1`] = `[]`;

exports[`loader should emit asset using the "messages" API: errors 1`] = `[]`;

exports[`loader should emit asset using the "messages" API: warnings 1`] = `[]`;

exports[`loader should emit warning using the "messages" API: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`loader should emit warning using the "messages" API: errors 1`] = `[]`;

exports[`loader should emit warning using the "messages" API: warnings 1`] = `
[
  "ModuleWarning: Module Warning (from \`replaced original path\`):
(10:3) from "postcss-plugin" plugin: <Message>

Code:
  color: green
",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
(14:3) from "postcss-plugin" plugin: <Message>

Code:
  color: blue
",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
(18:3) from "postcss-plugin" plugin: <Message>

Code:
  -x-border-color: blue blue *
",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
(19:3) from "postcss-plugin" plugin: <Message>

Code:
  -x-color: * #fafafa
",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
(23:3) from "postcss-plugin" plugin: <Message>

Code:
  -z-border-color: blue blue *
",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
(24:3) from "postcss-plugin" plugin: <Message>

Code:
  -z-color: * #fafafa
",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
(29:5) from "postcss-plugin" plugin: <Message>

Code:
  width: 500px
",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
(2:3) from "postcss-plugin" plugin: <Message>

Code:
  color: black
",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
(32:7) from "postcss-plugin" plugin: <Message>

Code:
  width: auto
",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
(36:7) from "postcss-plugin" plugin: <Message>

Code:
  color: white
",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
(41:5) from "postcss-plugin" plugin: <Message>

Code:
  display: block
",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
(6:3) from "postcss-plugin" plugin: <Message>

Code:
  color: red
",
]
`;

exports[`loader should register dependencies using the "messages" API: errors 1`] = `[]`;

exports[`loader should register dependencies using the "messages" API: warnings 1`] = `[]`;

exports[`loader should reuse PostCSS AST: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`loader should reuse PostCSS AST: errors 1`] = `[]`;

exports[`loader should reuse PostCSS AST: warnings 1`] = `[]`;

exports[`loader should throw an error on invalid syntax: errors 1`] = `
[
  "ModuleBuildError: Module build failed (from \`replaced original path\`):

SyntaxError

(1:3) /test/fixtures/css/style.css Unnecessary curly bracket

> 1 | a {
    |   ^
  2 |   color: black;
  3 | }
",
]
`;

exports[`loader should throw an error on invalid syntax: warnings 1`] = `[]`;

exports[`loader should work with SugarSS: css 1`] = `
"a {
  color: black
}
"
`;

exports[`loader should work with SugarSS: errors 1`] = `[]`;

exports[`loader should work with SugarSS: warnings 1`] = `[]`;

exports[`loader should work: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`loader should work: errors 1`] = `[]`;

exports[`loader should work: warnings 1`] = `[]`;
